docs/RFCS/20222202_historical_hot_ranges.md

```diff
@@ -0,0 +1,460 @@
+per range * 1000 ranges per sample * 4 samples per hour * 24 hours * 14 days =
+~363 Megabytes ```
+
+
+It's important to point out that the client will never need the majority of
+these bytes to be sent over the wire in order to render the visualization.
+Instead, the client will request samples in chunks, with subsequent API calls
+made to request the specific details of a range only as needed. This is
+discussed in more detail later on.
+
+### Metric Generation 
+
+The metric generation scheme is designed for both single and multi-tenant
+deployments, and operates with two different sampling modes: 
+#### Tenant Sampling 
+Tenant sampling generates metrics for a specific tenant. Tenant
+sampling is performed by the host, and metrics are stored such that a tenant can
+only access metrics for ranges inside its own keyspace. This prevents a tenant from
+accessing hotness from system-level ranges or another tenant's ranges. Every
+tenant receives its own 1000 range budget to prevent colder tenants from being
+overshadowed by hotter tenants.
+
+#### Host Sampling 
+Host sampling generates metrics reported over the cluster's
+entire keyspace in a tenant-agnostic way. The host receives its own 1000 range
+budget.
+
+A dedicated single-tenant deployment only performs host sampling, while
+multi-tenant deployments perform both tenant and host sampling. In a
+single-tenant deployment, the results of tenant's sampling sampling would be
+contained in the host sampling, and therefore be redundant.
+
+Tenant sampling and Host sampling follow mostly the same process - the
+differences manifest in how aggregation is performed to keep samples within
+budget, and how the sampled metrics are written to `system.hot_ranges`.
+
+According to the sample rate, a job is executed to
+request the 128 hottest replicas from every store via `replRankings`. These hot
+replicas are then sorted alphabetically by their start keys. If tenant sampling
+is being performed, each set of a tenant's reported replicas is
+limited to 1000 by a range reduction and aggregation scheme. This range
+reduction scheme is discussed in the next section. If host sampling is being
+performed, the global set of reported replicas is limited to 1000 by the same
+range reduction scheme. 
+
+After ranges have been aggregated and reduced, they are written to
+`system.hot_ranges`. Tenant sampling creates a row per tenant per sample, while
+Host sampling creates a single row per sample, whose `tenant_id` value is set to
+`'host'`. To query `system.hot_ranges`, demateralized views are created for each
+tenant and for the host.
+
+Finally, existing samples older than 2 weeks will be deleted.
+
+
+
+```sql 
+-- schema for system.hot_ranges
+CREATE TABLE system.hot_ranges ( 
+    sample_time INT NOT NULL,
+    
+    -- When host sampling is being performed, 
+    -- tenant_id will be set to 'host'
+    tenant_id INT NOT NULL,
+
+    -- info is defined below by `HHRSample`
+    info BYTES NOT NULL,
```

Comment:
I'll get back to you on this.

Reply:
> Can you help motivate the decision to use opaque protobufs in the storage here as opposed to sql columns?

It feels easier to evolve a protobuf definition over time than evolving a system table's schema. I also imagined at some level we would ship off protobufs to the front-end code rendering the chart -- I imagined shipping off these samples outright. For prefix compression, wouldn't it be easier to define it in protobuf form? Strawman:

```protobuf
message Sample {
  bytes shared_prefix = 1;
  message SpanStats {
    bytes start_key = 1;
    bytes end_key = 2;
    // stuff ...
  }
  repeated SpanStats span_stats = 2;
}
```

Is there a reason SQL columns are better? If the concern is around writing large binary blobs (ala system.jobs), we should give careful thought to what we include in our proto schema and also ensure that we're not rewriting things often.

Reply:
SQL columns are better because it's both easier and more efficient to query and aggregate them from SQL. I've noticed we've been having a bad time in statement stats largely because of the need to translate to and from json. Protobufs are even worse. Eliminating the possibility of the database performing computation on your behalf is almost certainly a mistake. There's a reason we don't sell a nosql database. 

All of this is to say that I've got my head in a space where in the fullness of time this would not be a system table and would not be beholden to the severe compatibility burden of the database internals. 

>  I also imagined at some level we would ship off protobufs to the front-end code rendering the chart -- I imagined shipping off these samples outright.

I think this is an anti-pattern. Sharing the durable data model on disk with the UI code leads to an impedance mismatch and hinders velocity. Pretty soon you forget that you don't need them to align. It's worth a thin translation layer up front to not incur tight coupling.

Reply:
I was also imagining what's stored in this table is already pre-aggregated. Is that not the case? It's already a cluster wide sample of what span statistics are over some time period, at which point the client code/thin translation layer would just read off of it with timestamps of interest (timestamps that do appear as columns for efficient access). If we envision other access queries (I would find it easier to know what they are if we pinned down the interface the client code needs from the backend), and they'd benefit from additional indexing, certainly we should add them as columns.

Reply:
If we stored individual samples as rows, over a two week period where each sample is over 15m, and say we maintain a fixed number of 1000 buckets, that's ((24*60*14)/15) * 1000 = 1,344,000 rows. I figured that'd be far too many for comfort, but maybe that's just FUD on my part.

Reply:
Is it reasonable to say that we can generate the schema as proposed, fill it 2 weeks worth of fake data, and see how it performs? (This is the current plan the kv-observability team is working on cc @koorosh)

Reply:
One thing I'm struggling with is how you'll render the keys given the boundaries move at each time step.

Reply:
sure. @ajwerner I'm going to move this to another thread.

Reply:
>I figured that'd be far too many for comfort, but maybe that's just FUD on my part.

With the rows each having keys in them, I'm definitely worried about it. For 1M rows of just integers, I think it'd be okay. It'll be better when we get the column projection scans @RaduBerinde has been toying with. 

I can also imagine some amount of normalization here where we map the values for each fixed bucket to then refer to some bucket metadata which lives longer than a bucket. By re-defining the buckets at each time step, we lose the ability to compress the metadata along that dimension. My sense is that that hurts from a data size perspective.

Reply:
I don’t have strong feelings on what the table layout should be. I think we’d benefit from defining first the interface we need the backend to provide, and then more specifically the storage layer powering it; both from the client code being powered by its contents, but also the process that’s aggregating stats cluster wide and persisting it. Whether the storage is backed by a single table with protos, multiple without, or not even using sql tables in the future, are all things interfaces would add color to. It’s not clear that we’ve settled on the interfaces just yet.